Deflake offset for leader epoch

joe-redpanda opened the pull request:
Describe paritition can occasionally return -1 aka UNKNOWN_EPOCH for a
partition's term.

This epoch is subsequently fed into offset_for_leader_epoch.

This is an illegal value for offset_for_leader_epoch.

This commit changes the logic for gathering partition data to require
first that all parition descriptions returned have a valid high
watermark and a valid epoch.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [x] v25.2.x
- [x] v25.1.x
- [x] v24.3.x

## Release Notes
### Bug Fixes

* deflake offset for leader epoch

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

joe-redpanda commented on src/v/kafka/server/handlers/offset_for_leader_epoch.cc:
```diff
@@ -56,33 +56,60 @@ get_epoch_end_offset(kafka::leader_epoch epoch, const partition_proxy& p) {
+            = +[](
+                 request_context& ctx,
+                 std::vector<epoch_end_offset>& ret,
+                 ntp_last_offset_request r) -> ss::future<void> {
```
delightfully this will be fixed in a future cpp standard

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75708-->
<!--e3bf7d3f-24fc-4ed2-8892-151c1f6034a4-->
<!--c691a70c-372f-420d-916d-41a4234630a3-->
<!--title-->
## Retry command for [Build#75708](https://buildkite.com/redpanda/redpanda/builds/75708)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/offset_for_leader_epoch_test.py::OffsetForLeaderEpochTest.test_offset_for_leader_epoch
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--cc631b25-90e2-4b0c-b644-6eec2cdd0708-->
<!--TEST RESULTS-->
<!--a9f50f11-27c7-4a69-934c-16a52ebca9dc-->
<!--TEST RESULTS-->
<!--3fe3550e-283e-4ac0-b2d6-ad94e71f504c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75708">build#75708</a></summary>


| test_class                      | test_method                    | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:--------------------------------|:-------------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ReplicatedMetastoreTest         | TestBasicRemoveTopics          |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/75708#019a563c-2934-44ee-9734-7d0cad1835e0 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                            |
| ShadowLinkingReplicationTests   | test_replication_with_failures | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75708#019a5659-9820-4270-a89b-ae493d71b826 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| OffsetForLeaderEpochTest        | test_offset_for_leader_epoch   | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75708#019a5659-9820-4270-a89b-ae493d71b826 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetForLeaderEpochTest&test_method=test_offset_for_leader_epoch        |
| OffsetForLeaderEpochTest        | test_offset_for_leader_epoch   | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75708#019a56bb-7563-4818-b52b-7219f9416965 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetForLeaderEpochTest&test_method=test_offset_for_leader_epoch        |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test       | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75708#019a56bb-7564-4481-80f7-620f9dacad31 | FLAKY         | 14/21    | upstream reliability is '97.25190839694656'. current run reliability is '66.66666666666666'. drift is 30.58524 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test     |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test       | {"cloud_storage_type": 1, "mixed_versions": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/75708#019a56bb-7565-4bb2-a1b1-090084d2dd09 | FLAKY         | 10/21    | upstream reliability is '95.4271961492178'. current run reliability is '47.61904761904761'. drift is 47.80815 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test     |
| ShadowLinkingRandomOpsTest      | test_node_operations           | {"failures": false}                                | integration | https://buildkite.com/redpanda/redpanda/builds/75708#019a56bb-7564-4481-80f7-620f9dacad31 | FLAKY         | 19/21    | upstream reliability is '98.94551845342706'. current run reliability is '90.47619047619048'. drift is 8.46933 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75830">build#75830</a></summary>


| test_class                      | test_method              | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                           |
|:--------------------------------|:-------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests   | test_topic_delete        | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75830#019a5efd-e175-492f-a285-e4e88277b42f | FLAKY         | 20/21    | upstream reliability is '98.125'. current run reliability is '95.23809523809523'. drift is 2.8869 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete          |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": true}                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75830#019a5f02-0be0-44af-ba71-83e6bc1b9bfb | FLAKY         | 19/21    | upstream reliability is '99.25558312655087'. current run reliability is '90.47619047619048'. drift is 8.77939 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| ScalingUpTest                   | test_fast_node_addition  | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75830#019a5f02-0bde-4749-a583-3e998d09231a | FLAKY         | 20/21    | upstream reliability is '97.93510324483776'. current run reliability is '95.23809523809523'. drift is 2.69701 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition                    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76764">build#76764</a></summary>


| test_class                           | test_method                | test_arguments                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                  |
|:-------------------------------------|:---------------------------|:-------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/76764#019aa402-28bc-481f-9f13-6d9250ca30a0 | FLAKY         | 20/21    | upstream reliability is '99.36842105263159'. current run reliability is '95.23809523809523'. drift is 4.13033 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| TopicRecoveryTest                    | test_many_partitions       | {"check_mode": "check_manifest_and_segment_metadata", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/76764#019aa402-28c1-4fe7-9a17-762dd473d5ce | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TopicRecoveryTest&test_method=test_many_partitions                          |
| WriteCachingFailureInjectionE2ETest  | test_crash_all             | {"use_transactions": false}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/76764#019aa3fa-ecdf-42f1-8c8d-0a5e98ea6acf | FLAKY         | 20/21    | upstream reliability is '90.927624872579'. current run reliability is '95.23809523809523'. drift is -4.31047 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all              |
</details>
<!--/body-->

mmaslankaprv commented on src/v/kafka/server/handlers/offset_for_leader_epoch.cc:
```diff
@@ -62,13 +62,27 @@ static ss::future<std::vector<epoch_end_offset>> fetch_offsets(
+                continue;
+            }
+        }
+
```
I am not certain this is really needed. It should be enough for us to check `is_leader` but do it in the same continuation as returning leader epoch

joe-redpanda replied:
is_leader can be stale though

As I understand this test:
publish data
inc epoch by leadership transfer
repeat N times

so the final request is asking for offsets for epoch N

is_leader can be stale, it **may** return true even though the current leader has lost leadership,
so it might return 'true' for a stale leader of epoch 'N-1', to which this broker responds to the request with epoch '-1' even though there is a real epoch N elsewhere.

so linearizable barrier corrects the problem by guaranteeing that this node is a non-stale leader at the point at which it attempts to handle the request.

bharathv replied:
> is_leader can be stale, it may return true even though the current leader has lost leadership,
so it might return 'true' for a stale leader of epoch 'N-1', to which this broker responds to the request with epoch '-1' even though there is a real epoch N elsewhere.

Can you help me clarify, in this case the code should return "unknown_leader_epoch", right? Because the requested_epoch (N) is > local stale epoch of `N-1` and the client has to deal with this error? no?

mmaslankaprv commented:
/ci-repeat 4
skip-redpanda-build
dt-repeat=10
tests/rptest/tests/offset_for_leader_epoch_test.py

joe-redpanda committed:
offset_for_leader_epoch_test.py: disallow epoch=-1

Describe paritition can occasionally return -1 aka UNKNOWN_EPOCH for a
partition's term.

This epoch is subsequently fed into offset_for_leader_epoch.

This is an illegal value for offset_for_leader_epoch.

This commit changes the logic for gathering partition data to require
first that all parition descriptions returned have a valid high
watermark and a valid epoch.

joe-redpanda commented:
I researched this again and found the following:

[DEBUG - 2025-10-29 16:51:31,697 - offset_for_leader_epoch_test - test_offset_for_leader_epoch_transfer - lineno:274]: Fetched offsets for epoch -1 : {... 30: 0, ...}, expected: {30: 612}

relevant invocation of offfset_for_leader_epoch:
BROKER  TOPIC             PARTITION  LEADER EPOCH  END OFFSET  ERROR
1       topic-kggcsntgbk  0          -1            0           
1       topic-kggcsntgbk  1          -1            0           
1       topic-kggcsntgbk  2          -1            0           
1       topic-kggcsntgbk  5          -1            0           
1       topic-kggcsntgbk  10         -1            0           
1       topic-kggcsntgbk  12         -1            0           
1       topic-kggcsntgbk  13         -1            0        


where did we get -1 for the epoch?

... /redpanda/redpanda/vbuild/redpanda_installs/ci/bin/rpk', 'topic', '-X', 'brokers=docker-rp-7:9092,docker-rp-22:9092,docker-rp-17:9092,docker-rp-23:9092,docker-rp-18:9092', 'describe', 'topic-kggcsntgbk', '-p', '-X', 'globals.request_timeout_overhead=30s', '-v']
[DEBUG - 2025-10-29 16:50:59,790 - rpk - _execute - lineno:1473]: 
PARTITION  LEADER  EPOCH  REPLICAS  LOG-START-OFFSET  HIGH-WATERMARK
0          2       4      [1 2 5]   0                 661
1          4       4      [1 4 5]   0                 617
2          3       4      [1 3 4]   0                 612
...
30         1       -1     [1 2 4]   0                 612   
...

RPK returned -1 for describe topic which was fed into offset_for_leader_epoch, which is an invalid value to provide, this results in an expectation mismatch between the offset 612, and the default offset which gets returned for an invalid partition which is the log_start offset

Should we check the provided epoch to guarantee that it is a valid epoch?

mmaslankaprv commented:
@joe-redpanda can we wait in the test for valid set of epochs ?

joe-redpanda commented:
> @joe-redpanda can we wait in the test for valid set of epochs ?

That's whats happening.

```python
    def _get_offsets_and_epochs(self, rpk: RpkTool, topic_name: str):
        offsets = []

        def refresh():
            result = rpk.describe_topic(topic_name)
            offsets.clear()
            offsets.extend(result)

        def all_offsets_valid():
            refresh()
            # metadata request may return INVALID_EPOCH aka -1
            # this should not be used because INVALID_EPOCH maps to latest available
            # epoch in OffsetForLeaderEpochRequest
            return all([p.high_watermark >= 0 and p.leader_epoch >= 0 for p in offsets])

        wait_until(all_offsets_valid, 30, 1)

        return offsets
```

we're doing a 30s wait until the condition is met where the condition is
```
p.high_watermark >= 0 and p.leader_epoch
```

joe-redpanda commented:
A good question, though, is should we strike down a user request with 'invalid_request' when epoch is < 0

bharathv commented:
> > @joe-redpanda can we wait in the test for valid set of epochs ?
> 
> That's whats happening.
> 
> ```python
>     def _get_offsets_and_epochs(self, rpk: RpkTool, topic_name: str):
>         offsets = []
> 
>         def refresh():
>             result = rpk.describe_topic(topic_name)
>             offsets.clear()
>             offsets.extend(result)
> 
>         def all_offsets_valid():
>             refresh()
>             # metadata request may return INVALID_EPOCH aka -1
>             # this should not be used because INVALID_EPOCH maps to latest available
>             # epoch in OffsetForLeaderEpochRequest
>             return all([p.high_watermark >= 0 and p.leader_epoch >= 0 for p in offsets])
> 
>         wait_until(all_offsets_valid, 30, 1)
> 
>         return offsets
> ```
> 
> we're doing a 30s wait until the condition is met where the condition is
> 
> ```
> p.high_watermark >= 0 and p.leader_epoch
> ```

I"m curious why epoch is -1 despite so many retries?

joe-redpanda commented:
> > > @joe-redpanda can we wait in the test for valid set of epochs ?
> > 
> > 
> > That's whats happening.
> > ```python
> >     def _get_offsets_and_epochs(self, rpk: RpkTool, topic_name: str):
> >         offsets = []
> > 
> >         def refresh():
> >             result = rpk.describe_topic(topic_name)
> >             offsets.clear()
> >             offsets.extend(result)
> > 
> >         def all_offsets_valid():
> >             refresh()
> >             # metadata request may return INVALID_EPOCH aka -1
> >             # this should not be used because INVALID_EPOCH maps to latest available
> >             # epoch in OffsetForLeaderEpochRequest
> >             return all([p.high_watermark >= 0 and p.leader_epoch >= 0 for p in offsets])
> > 
> >         wait_until(all_offsets_valid, 30, 1)
> > 
> >         return offsets
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > we're doing a 30s wait until the condition is met where the condition is
> > ```
> > p.high_watermark >= 0 and p.leader_epoch
> > ```
> 
> I"m curious why epoch is -1 despite so many retries?

previously the criteria was only p.high_watermark >= 0, with no epoch check

heres the old 'all' command

```python
          def all_offsets_valid():
                refresh()
                return all([p.high_watermark >= 0 for p in offsets])
```

I bumped this to a local function because both tests are susceptible to this.

vbotbuildovich commented:
/backport v25.3.x

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x
